src/java/com/puppetlabs/http/client/impl/Metrics.java

```diff
@@ -0,0 +1,257 @@
+package com.puppetlabs.http.client.impl;
+
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import org.apache.http.HttpRequest;
+import org.apache.http.RequestLine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class Metrics {
+    public static final String METRIC_NAMESPACE = "puppetlabs.http-client.experimental";
+    public static final String URL_NAMESPACE = "with-url";
+    public static final String URL_METHOD_NAMESPACE = "with-url-and-method";
+    public static final String ID_NAMESPACE = "with-metric-id";
+    public static final String BYTES_READ_STRING = "bytes-read";
+    public enum MetricType { BYTES_READ; }
+
+    public static String metricTypeString(Metrics.MetricType metricType) {
+        // currently this is the only metric type we have; in the future when
+        // there are multiple types this will do something more useful
+        return BYTES_READ_STRING;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Metrics.class);
+
+    synchronized private static ClientTimer getOrAddTimer(MetricRegistry metricRegistry,
+                                                          String name, ClientTimer newTimer) {
+        Metric timer = metricRegistry.getMetrics().get(name);
+
+        if ( timer == null ) {
+            return metricRegistry.register(name, newTimer);
```

Comment:
Should we do something more like what the `MetricRegistry.getOrAdd` call does that you mentioned?  They have a try/catch block around this part that would allow for the metric to still be returned rather than throwing an `IllegalArgumentException` if a metric with the same name and class type happens to be registered right around when this is called.  Not sure exactly why we'd want to deviate from their implementation for what we're doing?